### ISSUE TITLE
Use base index scale to access address in unsafe

### ISSUE THREAD

**Issue by fo40225** [dotnet/coreclr#2594]
Consider those code

```
int[] a = { 1, 2 };
int temp;
fixed (int* ptrA = a)
{
    for (int i = 0; i < a.Length; i++)
    {
        temp = ptrA[i];
        temp = temp + i;
        Console.WriteLine(temp.ToString());
    }

    int* tempPtrA = ptrA;
    for (int i = 0; i < a.Length; i++, tempPtrA++)
    {
        temp = *tempPtrA;
        temp = temp + i;
        Console.WriteLine(temp.ToString());
    }
}
```

In C/C++, first loop will faster than second one because first loop will use

```
mov         eax,dword ptr a[esi*4]
```

use only one instruction and get value. and second loop will use extra two instructions to increase pointer

```
mov         eax,dword ptr [edi]
inc         esi 
lea         edi,[edi+4]
```

But in C# unsafe code, first loop will use

```
mov         rcx,qword ptr [rbp-20h]
mov         qword ptr [rbp-38h],rcx
mov         rcx,qword ptr [rbp-38h]
mov         eax,dword ptr [rbp-2Ch]
movsxd      rax,eax  
mov         edx,4  
movsxd      rdx,edx  
imul        rax,rdx  
mov         ecx,dword ptr [rcx+rax]
```

and cause first one slow than second one. The `imul` slow down the performance.

If there is no performance penalty for using a scaling factor, why don't just use that instead of imul.

or there have some reason so can't use it?

**Comment by mikedn** [comment]
> In C/C++, first loop will faster than second one because first loop will use

That really depends on the compiler. Some may transform the slow version into the fast version and, some may transform the fast version into the slow version and some may do no transformation at all.

> But in C# unsafe code, first loop will use

I cannot reproduce that, it looks like debug code. The code I get for that loop is:

```
G_M11821_IG06:
       488B442420           mov      rax, bword ptr [rsp+20H]
       4863D7               movsxd   edx, edi
       8B1C90               mov      ebx, dword ptr [rax+4*rdx]
       03DF                 add      ebx, edi
; Console.WriteLine stuff omitted for brevity
       FFC7                 inc      edi
       3BF7                 cmp      esi, edi
       7FD4                 jg       SHORT G_M11821_IG06
```

It looks reasonable. The reload of `ptrA` at the start of the loop could have been avoided, this is caused by `fixed` and has already been reported in #2480. The `movsxd` could have been avoided by widening the `i` to long, this has been mentioned a few times in other JIT issues. But there's no `imul` to be seen in this code.

**Comment by fo40225** [comment]
OK, I try to use attach visual studio debugger after launch program instead of use visual studio to debug Release build directly. And I got your result.

So the problem is array style has unnecessary `movsxd` cause slower the pointer arithmetic?

the `ptrB[i] = ptrA[i];` loop will slower than `*tb = *ta;` in below code

    var a = new int[536870911];
    var rnd = new Random();

    for (int i = 0; i < a.Length; i++)
    {
        a[i] = rnd.Next();
    }

    var b = new int[536870911];
    fixed (int* ptrA = a, ptrB = b)
    {
        for (int i = 0; i < a.Length; i++)
        {
            ptrB[i] = ptrA[i];
        }
    }

    b = new int[536870911];
    fixed (int* ptrA = a, ptrB = b)
    {
        int* ta = ptrA;
        int* tb = ptrB;
        for (int i = 0; i < a.Length; i++, ta++, tb++)
        {
            *tb = *ta;
        }
    }

**Comment by mikedn** [comment]
> So the problem is array style has unnecessary  movsxd  cause slower the pointer arithmetic?

Have you measured and determined that it is indeed slower? For arrays that large the impact of memory loads and stores is significant enough that couple of additional fast instructions like `movsxd` or `lea  edi,[edi+4]` won't make much of a difference. Even a non-pointer version of this loop will provide the same performance, 365-410ms on my machine.

> the  ptrB[i] = ptrA[i];  loop will slower than  *tb = *ta;  in below code

That fact that the second loop in your code avoid using the fixed pointers has more impact than the use of fixed arithmetic. You can change the first loop to:

```C#
int* ta = ptrA;
int* tb = ptrB;
for (int i = 0; i < a.Length; i++)
    tb[i] = ta[i];
```
and the result won't be very different from the second loop, at least with the giant array you are testing with.

**Comment by fo40225** [comment]
I try the small array and there is no significant difference between non-fixed pointer array style and star style.

With increase array size, fixed pointer will have observably slow.

In very large array, fixed pointer will slow than non-fixed. even managed access will faster than fixed one.

I have no idea because those x86 asm looks very likely.

Why fixed pointer has access performance impact?

test code

    var x = new int[0x7FEFFFFF];
    for (int i = 0; i < x.Length; i++)
    {
        x[i] = rnd.Next();
    }

    var y = new int[0x7FEFFFFF];
    for (int i = 0; i < x.Length; i++)
    {
        y[i] = x[i];
    }

    y = new int[0x7FEFFFFF];
    fixed (int* ptrX = x, ptrY = y)
    {
        for (int i = 0; i < x.Length; i++)
        {
            ptrY[i] = ptrX[i];
        }
    }

    y = new int[0x7FEFFFFF];
    fixed (int* ptrX = x, ptrY = y)
    {
        int* tx = ptrX;
        int* ty = ptrY;
        for (int i = 0; i < a.Length; i++)
        {
            ty[i] = tx[i];
        }
    }

    y = new int[0x7FEFFFFF];
    fixed (int* ptrX = x, ptrY = y)
    {
        int* tx = ptrX;
        int* ty = ptrY;
        for (int i = 0; i < a.Length; i++, tx++, ty++)
        {
            *ty = *tx;
        }
    }

**Comment by fo40225** [comment]
Oh I figure out, is that the same problem in # 2480?

**Comment by mikedn** [comment]
If I understand correctly what you are asking then yes, it's the same problem as in #2480.

Let's look at the generated code for those 2 loops that use indexing:

```C#
            fixed (int* ptrX = x, ptrY = y)
            {
                for (int i = 0; i < x.Length; i++)
                {
                    ptrY[i] = ptrX[i];
                }
            }
```
generates
```
G_M52619_IG09:
       4C8B442438           mov      r8, bword ptr [rsp+38H] ; reload ptrY
       4C63C9               movsxd   r9d, ecx
       4C8B542440           mov      r10, bword ptr [rsp+40H] ; reload ptrX
       478B148A             mov      r10d, dword ptr [r10+4*r9]
       47891488             mov      dword ptr [r8+4*r9], r10d
       FFC1                 inc      ecx
       3BD1                 cmp      edx, ecx
       7FE5                 jg       SHORT G_M52619_IG09
```
and
```C#
            fixed (int* ptrX = x, ptrY = y)
            {
                int* tx = ptrX;
                int* ty = ptrY;
                for (int i = 0; i < x.Length; i++)
                {
                    ty[i] = tx[i];
                }
            }
```
generates
```
G_M52619_IG17:
       4D63C8               movsxd   r9d, r8d
       468B1488             mov      r10d, dword ptr [rax+4*r9]
       46891489             mov      dword ptr [rcx+4*r9], r10d
       41FFC0               inc      r8d
       413BD0               cmp      edx, r8d
       7FED                 jg       SHORT G_M52619_IG17
```
Note how in the first version ptrX and ptrY are reloaded from local variables on every iteration while in the second version they are not.

I think it's also worth looking at the non-pointer version:
```C#
            for (int i = 0; i < x.Length; i++)
            {
                y[i] = x[i];
            }
```
generates
```
G_M52619_IG04:
       4863CA               movsxd   ecx, edx
       448B448E10           mov      r8d, dword ptr [rsi+4*rcx+16]
       4489448810           mov      dword ptr [rax+4*rcx+16], r8d
       FFC2                 inc      edx
       81FAFFFFEF7F         cmp      edx, 0x7FEFFFFF
       7CE9                 jl       SHORT G_M52619_IG04
```
Usually only one of the necessary array bound checks is eliminated but in this particular case the JIT managed to eliminate both checks because the array lengths are known. The generating code is identical to the one generated by one of the pointer based versions.

**Comment by fo40225** [comment]
Thanks for the explanation.

So in this access array scenario, managed code and unmanaged code didn't have performance difference.

Can we say if JIT can eliminated the bound checks, there is nothing difference between managed and unsafe code?

**Comment by mikedn** [comment]
> Can we say if JIT can eliminated the bound checks, there is nothing difference between managed and unsafe code?

Sort of. We've seen that both managed and unsafe versions can generate identical code so it would seem that the answer is simply yes. But in unsafe versions you may be able to do optimizations that aren't possible in the managed version.

In this particular case you could cast `int*` to `long*` and get a speed bump because it is faster to copy 8 bytes at a time instead of 4:

```C#
        [MethodImpl(MethodImplOptions.NoInlining)]
        private static void Test(int* px, int* py, long length)
        {
            long* lx = (long*)px;
            long* ly = (long*)py;

            for (long i = 0; i < length / 2; i++) // length / 2 assumes that length is a multiple of 2
                ly[i] = lx[i];
        }
```

These days you could achieve a similar effect without using unsafe code by using `System.Numerics.Vector<T>`. But before SIMD support was added to .NET there was no other way to do something like this, it isn't possible to cast `int[]` to `long[]` like you can cast `int*` to `long*`.

**Comment by Janiels** [comment]
> it isn't possible to cast int[] to long[] like you can cast int* to long* 

```
private static void CopyFast(int[] from, int[] to)
{
	long[] fromAsLongs = new Caster {Int32Array = from}.Int64Array;
	long[] toAsLongs = new Caster {Int32Array = to}.Int64Array;

	// fromAsLongs.Length will be wrong. But since it's always larger
	// (as sizeof(int) < sizeof(long)), generated bounds checks are still
	// happy.
	int numLongs = from.Length/2;
	for (int i = 0; i < numLongs; i++)
		toAsLongs[i] = fromAsLongs[i];
}

[StructLayout(LayoutKind.Explicit)]
private struct Caster
{
	[FieldOffset(0)]
	public int[] Int32Array;
	[FieldOffset(0)]
	public long[] Int64Array;
}
```
Of course this trick is very dirty and relies on the layout of arrays of all types being equal. And it won't work for larger -> smaller, as the bounds checks will begin failing. But assuming the JIT hoists the bounds check, it should see that `numLongs` < `toAsLongs.Length` and `fromAsLongs.Length`. I haven't actually checked the generated code, though.

**Comment by mikedn** [comment]
> Of course this trick is very dirty, unverifiable, type-safety breaking and relies on the layout of arrays of all types being equal.

:scream_cat:, this makes unsafe code look very safe. 

> But assuming the JIT hoists the bounds check, it should see that  numLongs  <  toAsLongs.Length  and  fromAsLongs.Length  and then it may be faster.

I wonder what this does to loop cloning. I don't think I want to find out.

**Comment by Janiels** [comment]
> I wonder what this does to loop cloning. I don't think I want to find out.
I didn't know the JIT actually applied this optimization. Then it may be able to eliminate all bounds checks inside the loop.
If you do check it, feel free to share the results! :smiley:

**Comment by mikedn** [comment]
> I didn't know the JIT actually applied this optimization. 

Yep, RyuJIT does this. Though I find the feature questionable due to the large amount of code it generates.

> Then it may be able to eliminate all bounds checks inside the loop.

I looked at the generated code (too large to include here, take a look at this [gist](https://gist.github.com/mikedn/503f4f121ae1fbcdf290) if you want) and it seems that it would work. It simply decides whether to use the loop version with bound checks or the one without bound checks. Either version would work fine, as if no loop cloning was done.

**Comment by Janiels** [comment]
>I looked at the generated code (too large to include here, take a look at this gist if you want) and it seems that it would work.

Yeah, that actually looks pretty good. Better than the equivalent source. :smile:

**Comment by mikedn** [comment]
> Yeah, that actually looks pretty good. Better than the equivalent source. 

Note that I extracted the for loop to a separate method and assembly dump is only from that method.

> Though the hoisted checks look a little convoluted, but I have not checked if those could really be improved. And given that it's pure registers, there probably are no big speed implications.

The null checks at the start seem useless. There also seems to be a redundant and possibly wrong check `r8 >= 0` in `G_M33861_IG03`.

**Comment by fo40225** [comment]
In fact, if I want to do a array copy, I will use `Array.Copy`, `Buffer.BlockCopy`, `Buffer.MemoryCopy`

Those method will faster than unsafe code and I don't know way. maybe they have some black magic optimize in there.

Cast int array to long to double the copy operation looks very tricky. If I must do that I will use the `System.Numerics.Vector<T>` to do that SIMD operation. The code will look more reasonable.

The original sample code just want to show access array, not specific array copy.

Thanks all, I learned that I can use long to access array 8 bytes at once.

The `struct Caster` is awesome, never think before `FieldOffset` can be abused to do that.

About the bounds checks, the optimization role is loop condition can be determined and use the index to access array will eliminate all bounds checks. Do I understand right?

**Comment by mikedn** [comment]
> In fact, if I want to do a array copy, I will use  Array.Copy ,  Buffer.BlockCopy ,  Buffer.MemoryCopy 
> Those method will faster than unsafe code and I don't know way. maybe they have some black magic optimize in there.

The source of MemoryCopy is [here](https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/Buffer.cs#L262). The usual black magic involved in copying is to copy as much as possible in a single instruction. For example, to copy 9 bytes `MemoryCopy` copies one `long` and one `byte`.

`Array.Copy` and `Buffer.BlockCopy` also have the source in this repository but ultimately they end up calling C's `memcpy` function. There's a long discussion about copying (including SIMD use) in #2430.

> About the bounds checks, the optimization role is loop condition can be determined and use the index to access array will eliminate all bounds checks. Do I understand right?

In short, if the loop condition has the form `i < a.Length` then bound checks for all accesses of the form `a[i]` are eliminated.

**Comment by Janiels** [comment]
> The null checks at the start seem useless. There also seems to be a redundant and possibly wrong check r8 >= 0 in G_M33861_IG03.

Hmm yes, that looks redundant (given that it has already been checked for the entire 64-bit register earlier). I don't think it's wrong though (I assume `r8` is the argument with value `from.Length / 2` here). Does the ABI not guarantee that the upper 32 bits are zero when passing a 32-bit value in a register?

The null checks could be because the JIT doesn't want to deal with throwing null refs for AVs inside the hoisted check (`G_M33861_IG03`). Not entire sure how the JIT tracks this information, however, so I'm just speculating.

**Comment by Janiels** [comment]
> About the bounds checks, the optimization role is loop condition can be determined and use the index to access array will eliminate all bounds checks. Do I understand right?

[Here](http://blogs.msdn.com/b/clrcodegeneration/archive/2009/08/13/array-bounds-check-elimination-in-the-clr.aspx) is a nice blog post about it.

**Comment by mikedn** [comment]
> Does the ABI not guarantee that the upper 32 bits are zero when passing a 32-bit value in a register?

Good question. Given the x64 specifics it's very likely that the upper 32 bits will be 0 but I don't know if that's guaranteed by the JIT.

> The null checks could be because the JIT doesn't want to deal with throwing null refs for AVs inside the hoisted check ( G_M33861_IG03 ). 

Yes but the hoisted check isn't executed if the copy length is 0 due to the r8d check at the start of the method. I'll take a look at the JIT code, maybe I'll end up opening a separate issue if this code is really useless.

**Comment by Janiels** [comment]
> Yes but the hoisted check isn't executed if the copy length is 0 due to the r8d check at the start of the method.

But is `r8d` not just another argument you have passed? The arrays could be null even if this is non-zero.
Maybe I'm just confused since I don't know exactly what C# code this has been generated for. Can you share the exact code?

**Comment by mikedn** [comment]
The C# code is
```C#
[MethodImpl(MethodImplOptions.NoInlining)]
private static void Copy(long[] src, long[] dst, int length)
{
    for (int i = 0; i < length; i++)
        dst[i] = src[i];
}
```
If length is non zero, arrays are null and the null is missing a NullReferenceException should be thrown from the hoisted bounds check and that's perfectly normal. 

Presumably you are concerned with the line number information that's included in the exception. The hoisted code should have the same IL offset as the original array accesses so the exception strack trace should point to the `dst[i] = src[i];` line even if the exception occurs in hoisted code. CoreCLR doesn't seem to show line number information in strack traces so I can't check but with .NET 4.6.1 and VS2015 the debugger points to the `i < length` loop condition. So even with those explicit null checks the IL mapping is still wrong.

**Comment by Janiels** [comment]
> If length is non zero, arrays are null and the null is missing a NullReferenceException should be thrown from the hoisted bounds check and that's perfectly normal.

This was the part I was speculating that perhaps they did not want to bother with. If this is the case, then I definitely agree with your analysis that the null checks are redundant.

**Comment by AndyAyersMS** [comment]
Seems like there is no actual issue here. Am I overlooking something?

**Comment by fo40225** [comment]
@AndyAyersMS 

After test, unsafe pointer still has performance issue.

Test code [here](https://github.com/fo40225/UnsafePerfTest/blob/master/UnsafePerfTest/Program.cs)

#### Result

.NET Core

    BenchmarkDotNet=v0.10.14, OS=Windows 10.0.15063.1029 (1703/CreatorsUpdate/Redstone2)
    Intel Xeon CPU E3-1230 V2 3.30GHz, 1 CPU, 8 logical and 4 physical cores
    Frequency=3222678 Hz, Resolution=310.3009 ns, Timer=TSC
    .NET Core SDK=2.1.300-rc1-008662
      [Host]     : .NET Core 2.1.0-rc1-26423-06 (CoreCLR 4.6.26423.02, CoreFX 4.6.26423.06), 64bit RyuJIT
      DefaultJob : .NET Core 2.1.0-rc1-26423-06 (CoreCLR 4.6.26423.02, CoreFX 4.6.26423.06), 64bit RyuJIT
    
    
                                   Method |     N |     Mean |     Error |    StdDev |
    ------------------------------------- |------ |---------:|----------:|----------:|
                       ManagedAccessArray | 10000 | 1.585 ms | 0.0329 ms | 0.0352 ms |
                  UnsafeAccessArrayOffset | 10000 | 1.428 ms | 0.0277 ms | 0.0296 ms |
         UnsafeAccessArrayWorkroundOffset | 10000 | 1.165 ms | 0.0229 ms | 0.0215 ms |
     UnsafeAccessArrayWorkroundRawPointer | 10000 | 1.166 ms | 0.0216 ms | 0.0191 ms |

.NET Framework

    BenchmarkDotNet=v0.10.14, OS=Windows 10.0.15063.1029 (1703/CreatorsUpdate/Redstone2)
    Intel Xeon CPU E3-1230 V2 3.30GHz, 1 CPU, 8 logical and 4 physical cores
    Frequency=3222678 Hz, Resolution=310.3009 ns, Timer=TSC
      [Host]     : .NET Framework 4.6.2 (CLR 4.0.30319.42000), 64bit RyuJIT-v4.7.2115.0
      DefaultJob : .NET Framework 4.6.2 (CLR 4.0.30319.42000), 64bit RyuJIT-v4.7.2115.0
    
    
                                   Method |     N |     Mean |     Error |    StdDev |   Median |
    ------------------------------------- |------ |---------:|----------:|----------:|---------:|
                       ManagedAccessArray | 10000 | 2.274 ms | 0.0175 ms | 0.0155 ms | 2.270 ms |
                  UnsafeAccessArrayOffset | 10000 | 2.198 ms | 0.0439 ms | 0.0916 ms | 2.159 ms |
         UnsafeAccessArrayWorkroundOffset | 10000 | 1.872 ms | 0.0253 ms | 0.0236 ms | 1.863 ms |
     UnsafeAccessArrayWorkroundRawPointer | 10000 | 1.876 ms | 0.0132 ms | 0.0110 ms | 1.878 ms |

**Comment by mikedn** [comment]
> After test, unsafe pointer still has performance issue.

Presumably you're referring to the case of `UnsafeAccessArrayOffset` that's slower than the other 2 unsafe versions?
```
          UnsafeAccessArrayOffset |      Core |    Core | 10000 | 1.468 ms | 0.0395 ms | 0.0369 ms |
 UnsafeAccessArrayOffsetWorkround |      Core |    Core | 10000 | 1.168 ms | 0.0228 ms | 0.0263 ms |
      UnsafeAccessArrayRawPointer |      Core |    Core | 10000 | 1.170 ms | 0.0233 ms | 0.0249 ms |

```

**Comment by mikedn** [comment]
With the current .NET Core build I get the following results:
```
               ManagedAccessArray | 10000 | 1.702 ms | 0.0188 ms | 0.0167 ms |
          UnsafeAccessArrayOffset | 10000 | 1.379 ms | 0.0082 ms | 0.0076 ms |
 UnsafeAccessArrayOffsetWorkround | 10000 | 1.404 ms | 0.0152 ms | 0.0127 ms |
      UnsafeAccessArrayRawPointer | 10000 | 1.388 ms | 0.0204 ms | 0.0160 ms |
```
So there's practically no difference between the 3 unsafe versions.
The first 2 unsafe versions produce identical loop code so they can't really produce different timings unless code alignment interferes. The last unsafe version produce slightly different code but the different is unlikely to affect performance.
```asm
; UnsafeAccessArrayOffset & UnsafeAccessArrayOffsetWorkround loop code
G_M33474_IG09:
       4963C0               movsxd   rax, r8d
       448B0C82             mov      r9d, dword ptr [rdx+4*rax]
       44890C81             mov      dword ptr [rcx+4*rax], r9d
       41FFC0               inc      r8d
       48B838296999E7010000 mov      rax, 0x1E799692938
       488B00               mov      rax, gword ptr [rax]
       44394008             cmp      dword ptr [rax+8], r8d
       7FDF                 jg       SHORT G_M33474_IG09
```

```asm
; UnsafeAccessArrayRawPointer loop code
G_M33474_IG09:
       8B02                 mov      eax, dword ptr [rdx]
       8901                 mov      dword ptr [rcx], eax
       41FFC0               inc      r8d
       4883C204             add      rdx, 4
       4883C104             add      rcx, 4
       48B83829703B19020000 mov      rax, 0x2193B702938
       488B00               mov      rax, gword ptr [rax]
       44394008             cmp      dword ptr [rax+8], r8d
       7FDE                 jg       SHORT G_M33474_IG09
```

**Comment by fo40225** [comment]
I use nightly build from https://github.com/dotnet/versions/tree/master/build-info/dotnet/product/cli/release/2.1 and get no luck.

    BenchmarkDotNet=v0.10.14, OS=Windows 10.0.15063.1029 (1703/CreatorsUpdate/Redstone2)
    Intel Xeon CPU E3-1230 V2 3.30GHz, 1 CPU, 8 logical and 4 physical cores
    Frequency=3222678 Hz, Resolution=310.3009 ns, Timer=TSC
    .NET Core SDK=2.1.300-rc1-008662
      [Host]     : .NET Core 2.1.0-rc1-26424-03 (CoreCLR 4.6.26424.05, CoreFX 4.6.26424.04), 64bit RyuJIT
      DefaultJob : .NET Core 2.1.0-rc1-26424-03 (CoreCLR 4.6.26424.05, CoreFX 4.6.26424.04), 64bit RyuJIT
    
    
                               Method |     N |     Mean |     Error |    StdDev |
    --------------------------------- |------ |---------:|----------:|----------:|
                   ManagedAccessArray | 10000 | 1.509 ms | 0.0131 ms | 0.0122 ms |
              UnsafeAccessArrayOffset | 10000 | 1.355 ms | 0.0137 ms | 0.0114 ms |
     UnsafeAccessArrayOffsetWorkround | 10000 | 1.113 ms | 0.0067 ms | 0.0059 ms |
          UnsafeAccessArrayRawPointer | 10000 | 1.118 ms | 0.0102 ms | 0.0085 ms |

Where should I get latest SDK installer? 

This page (https://github.com/dotnet/cli) only provide 26411-06.

**Comment by mikedn** [comment]
Can you try using BenchmarkDotNet's [disassembly diagnoser](http://adamsitnik.com/Disassembly-Diagnoser/) to get the disassembly for each test? We need to figure out if the differences are due to a different .NET Core version (unlikely, I'm not aware of any significant changes between the one you're using and the current master build) or due to different CPUs.

**Comment by fo40225** [comment]
``` ini

BenchmarkDotNet=v0.10.14, OS=Windows 10.0.15063.1029 (1703/CreatorsUpdate/Redstone2)
Intel Xeon CPU E3-1230 V2 3.30GHz, 1 CPU, 8 logical and 4 physical cores
Frequency=3222678 Hz, Resolution=310.3009 ns, Timer=TSC
.NET Core SDK=2.1.300-rc1-008662
  [Host]     : .NET Core 2.1.0-rc1-26424-03 (CoreCLR 4.6.26424.05, CoreFX 4.6.26424.04), 64bit RyuJIT
  DefaultJob : .NET Core 2.1.0-rc1-26424-03 (CoreCLR 4.6.26424.05, CoreFX 4.6.26424.04), 64bit RyuJIT


```
|                           Method |     N |     Mean |     Error |    StdDev |
|--------------------------------- |------ |---------:|----------:|----------:|
|               ManagedAccessArray | 10000 | 1.510 ms | 0.0089 ms | 0.0075 ms |
|          UnsafeAccessArrayOffset | 10000 | 1.383 ms | 0.0270 ms | 0.0379 ms |
| UnsafeAccessArrayOffsetWorkround | 10000 | 1.120 ms | 0.0091 ms | 0.0080 ms |
|      UnsafeAccessArrayRawPointer | 10000 | 1.121 ms | 0.0072 ms | 0.0060 ms |


``` asm
UnsafePerfTest.UnsafeTest.ManagedAccessArray()
mov     rcx,offset System_Private_CoreLib+0xa0f890
mov     edx,100000h
call    coreclr!MetaDataGetDispenser+0x2f690
xor     edx,edx
mov     rcx,13B26316D20h
mov     rcx,qword ptr [rcx]
cmp     dword ptr [rcx+8],0
jle     M00_L01
M00_L00
 
mov     rcx,13B26316D20h
mov     rcx,qword ptr [rcx]
cmp     edx,dword ptr [rcx+8]
jae     00007ff8`432c165d
movsxd  r8,edx
mov     ecx,dword ptr [rcx+r8*4+10h]
cmp     edx,100000h
jae     00007ff8`432c165d
mov     dword ptr [rax+r8*4+10h],ecx
inc     edx
mov     rcx,13B26316D20h
mov     rcx,qword ptr [rcx]
cmp     dword ptr [rcx+8],edx
jg      M00_L00
M00_L01
 
add     rsp,28h
 
UnsafePerfTest.UnsafeTest.UnsafeAccessArrayOffset()
mov     rcx,offset System_Private_CoreLib+0xa0f890
mov     edx,100000h
call    coreclr!MetaDataGetDispenser+0x2f690
mov     rdx,2A0B9106D20h
mov     rdx,qword ptr [rdx]
mov     rcx,rdx
test    rcx,rcx
je      M00_L00
mov     r8d,dword ptr [rcx+8]
test    r8d,r8d
jne     M00_L01
M00_L00
 
xor     r8d,r8d
mov     qword ptr [rsp+30h],r8
jmp     M00_L02
M00_L01
 
cmp     r8d,0
jbe     00007ff8`432d16a6
add     rcx,10h
mov     qword ptr [rsp+30h],rcx
M00_L02
 
mov     ecx,dword ptr [rax+8]
test    ecx,ecx
jne     M00_L03
xor     eax,eax
mov     qword ptr [rsp+28h],rax
jmp     M00_L04
M00_L03
 
add     rax,10h
mov     qword ptr [rsp+28h],rax
M00_L04
 
xor     eax,eax
cmp     dword ptr [rdx+8],0
jle     M00_L06
M00_L05
 
mov     rdx,qword ptr [rsp+28h]
movsxd  rcx,eax
mov     r8,qword ptr [rsp+30h]
mov     r8d,dword ptr [r8+rcx*4]
mov     dword ptr [rdx+rcx*4],r8d
inc     eax
mov     rdx,2A0B9106D20h
mov     rdx,qword ptr [rdx]
cmp     dword ptr [rdx+8],eax
jg      M00_L05
M00_L06
 
xor     eax,eax
mov     qword ptr [rsp+30h],rax
mov     qword ptr [rsp+28h],rax
add     rsp,38h
 
UnsafePerfTest.UnsafeTest.UnsafeAccessArrayOffsetWorkround()
mov     rcx,offset System_Private_CoreLib+0xa0f890
mov     edx,100000h
call    coreclr!MetaDataGetDispenser+0x2f690
mov     rdx,15510006D20h
mov     rdx,qword ptr [rdx]
mov     rcx,rdx
test    rcx,rcx
je      M00_L00
mov     r8d,dword ptr [rcx+8]
test    r8d,r8d
jne     M00_L01
M00_L00
 
xor     r8d,r8d
mov     qword ptr [rsp+30h],r8
jmp     M00_L02
M00_L01
 
cmp     r8d,0
jbe     00007ff8`432b16a9
add     rcx,10h
mov     qword ptr [rsp+30h],rcx
M00_L02
 
mov     ecx,dword ptr [rax+8]
test    ecx,ecx
jne     M00_L03
xor     eax,eax
mov     qword ptr [rsp+28h],rax
jmp     M00_L04
M00_L03
 
add     rax,10h
mov     qword ptr [rsp+28h],rax
M00_L04
 
mov     rax,qword ptr [rsp+30h]
mov     rcx,qword ptr [rsp+28h]
xor     r8d,r8d
cmp     dword ptr [rdx+8],0
jle     M00_L06
M00_L05
 
movsxd  rdx,r8d
mov     r9d,dword ptr [rax+rdx*4]
mov     dword ptr [rcx+rdx*4],r9d
inc     r8d
mov     rdx,15510006D20h
mov     rdx,qword ptr [rdx]
cmp     dword ptr [rdx+8],r8d
jg      M00_L05
M00_L06
 
xor     eax,eax
mov     qword ptr [rsp+30h],rax
mov     qword ptr [rsp+28h],rax
add     rsp,38h
 
UnsafePerfTest.UnsafeTest.UnsafeAccessArrayRawPointer()
mov     rcx,offset System_Private_CoreLib+0xa0f890
mov     edx,100000h
call    coreclr!MetaDataGetDispenser+0x2f690
mov     rdx,25E9D846D20h
mov     rdx,qword ptr [rdx]
mov     rcx,rdx
test    rcx,rcx
je      M00_L00
mov     r8d,dword ptr [rcx+8]
test    r8d,r8d
jne     M00_L01
M00_L00
 
xor     r8d,r8d
mov     qword ptr [rsp+30h],r8
jmp     M00_L02
M00_L01
 
cmp     r8d,0
jbe     00007ff8`432c16aa
add     rcx,10h
mov     qword ptr [rsp+30h],rcx
M00_L02
 
mov     ecx,dword ptr [rax+8]
test    ecx,ecx
jne     M00_L03
xor     eax,eax
mov     qword ptr [rsp+28h],rax
jmp     M00_L04
M00_L03
 
add     rax,10h
mov     qword ptr [rsp+28h],rax
M00_L04
 
mov     rax,qword ptr [rsp+30h]
mov     rcx,qword ptr [rsp+28h]
xor     r8d,r8d
cmp     dword ptr [rdx+8],0
jle     M00_L06
M00_L05
 
mov     edx,dword ptr [rax]
mov     dword ptr [rcx],edx
inc     r8d
add     rax,4
add     rcx,4
mov     rdx,25E9D846D20h
mov     rdx,qword ptr [rdx]
cmp     dword ptr [rdx+8],r8d
jg      M00_L05
M00_L06
 
xor     eax,eax
mov     qword ptr [rsp+30h],rax
mov     qword ptr [rsp+28h],rax
add     rsp,38h
 
```

**Comment by mikedn** [comment]
So your `UnsafeAccessArrayOffset` suffers from the issue with fixed variables:
```diff
--- C:\Users\Mike\Desktop\asm2.asm	2018-04-26 20:57:11.000000000 +-0300
+++ C:\Users\Mike\Desktop\asm1.asm	2018-04-26 20:57:18.000000000 +-0300
@@ -41,12 +43,10 @@
 M00_L05
  
-mov     rdx,qword ptr [rsp+28h]
-movsxd  rcx,eax
-mov     r8,qword ptr [rsp+30h]
-mov     r8d,dword ptr [r8+rcx*4]
-mov     dword ptr [rdx+rcx*4],r8d
-inc     eax
-mov     rdx,2A0B9106D20h
+movsxd  rdx,r8d
+mov     r9d,dword ptr [rax+rdx*4]
+mov     dword ptr [rcx+rdx*4],r9d
+inc     r8d
+mov     rdx,15510006D20h
 mov     rdx,qword ptr [rdx]
-cmp     dword ptr [rdx+8],eax
+cmp     dword ptr [rdx+8],r8d
 jg      M00_L05
```
@AndyAyersMS AFAIR you fixed this, can't find the PR to link to

**Comment by fo40225** [comment]
```
1>    C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\15.0\Bin\Roslyn\csc.exe /noconfig /unsafe+ /checked-
1>    搭配來自目錄的編譯器使用共用編譯: C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\15.0\Bin\Roslyn
1>    Microsoft (R) Visual C# 編譯器 版 2.4.0.62122 (ab56a4a6)
1>    Copyright (C) Microsoft Corporation. 著作權所有，並保留一切權利。
1>    輸出項目: CscCommandLineArgs=
1>  工作 "Csc" 執行完成。
```

Roslyn 2.6.0 has fixed this issue.
